### PR TITLE
[TPU] Support PyTorch/XLA FSDP via SPMD

### PR DESCRIPTION
**Pull request by alanwaketan**
# What does this PR do?

Summary:
This is the first attempt to enable FSDP via SPMD (FSDPv2) on PyTorch/XLA model.

More information about FSDPv2 can be found here:
1. A user guide: https://github.com/pytorch/xla/blob/master/docs/fsdpv2.md
2. A RFC: https://github.com/pytorch/xla/issues/6379

Besides the initial implementation of FSDPv2 in r2.2, this change will also requires the following changes in PyTorch/XLA:
1. https://github.com/pytorch/xla/pull/6499
2. https://github.com/pytorch/xla/pull/6500
3. https://github.com/pytorch/xla/pull/6498
4. https://github.com/pytorch/xla/pull/6525
Therefore, it will only be compatible with the nightly builds.

Example use cases:
1. Prepare a FSDPv2 config:
```json
{
    "fsdp_transformer_layer_cls_to_wrap": [
        "LlamaDecoderLayer"
    ],
    "xla": true,
    "xla_fsdp_v2": true,
    "xla_fsdp_grad_ckpt": true
}
```
2. Invoke the trainer using the following command:
```
XLA_USE_SPMD=1 XLA_USE_BF16=1 python3 examples/pytorch/language-modeling/run_clm.py     --num_train_epochs 1     --dataset_name wikitext     --dataset_config_name wikitext-2-raw-v1 --per_device_train_batch_size 128     --do_train     --output_dir /tmp/test-clm     --overwrite_output_dir     --config_name ../transformers_pt/2B.config     --cache_dir /tmp     --tokenizer_name hf-internal-testing/llama-tokenizer      --block_size 1024     --optim adafactor     --save_strategy no     --logging_strategy no --fsdp "full_shard" --fsdp_config fsdp_config.json --torch_dtype bfloat16 --dataloader_drop_last yes
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @younesbelkada

### CODE REVIEW

**Comment by alanwaketan** [comment]
Can HF folks point me on how to add test case in this case and also how to update the documentation?

**Comment by alanwaketan** [comment]
cc @yeounoh @jonb377

**Review by ArthurZucker — APPROVED** [review]
LGTM overall! We might want to add a small test, it can be done in a followup PR. 
Pinging @muellerzr for a second look!

**Comment by ArthurZucker** [comment]
Tests should be added in the `tests/trainer/test_trainer.py` file. You should find similar tests!

**Review by muellerzr — COMMENTED** [review]
As @ArthurZucker hinted at, we now don't handle things like this in the trainer directly. I would rather see this code over in accelerate which we can then bring into Trainer automatically since it relies on it for preparation. Especially as this deals with the dataloaders. Would that be possible please! :)

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1380,6 +1390,11 @@ def _wrap_model(self, model, training=True, dataloader=None):
                     size_based_auto_wrap_policy,
                     transformer_auto_wrap_policy,
                 )
+
+                if self.is_fsdp_xla_v2_enabled:
+                    from torch_xla.experimental.spmd_fully_sharded_data_parallel import (
+                        SpmdFullyShardedDataParallel as FSDPv2,
```
Comment (muellerzr) [inline src/transformers/trainer.py:1396]: Could we make this easier by importing `FSDPv2` as `FSDP` instead?

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -1411,15 +1426,42 @@ def _wrap_model(self, model, training=True, dataloader=None):
...
+                    elif isinstance(output, tuple):
+                        real_output = output[0]
+                    elif isinstance(output, CausalLMOutputWithPast):
+                        real_output = output.logits
+
+                    if real_output is None:
+                        raise ValueError("Something went wrong, the output of the model shouldn't be `None`")
+                    xs.mark_sharding(real_output, mesh, ("fsdp", None, None))
+
+                self.model = model = FSDPv2(
```
Comment (muellerzr) [inline src/transformers/trainer.py:1452]: And then leave the check for down here on what to do.

**Comment by alanwaketan** [comment]
> As @ArthurZucker hinted at, we now don't handle things like this in the trainer directly. I would rather see this code over in accelerate which we can then bring into Trainer automatically since it relies on it for preparation. Especially as this deals with the dataloaders. Would that be possible please! :)

Can you elaborate it a bit more? I can move the `model = model.to(xm.xla_device())` logic. But for the dataloader logic, i.e., tpu_spmd_dataloader, where do you suggest me to move it to?

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -1411,15 +1426,42 @@ def _wrap_model(self, model, training=True, dataloader=None):
...
             # Wrap the base model with an outer FSDP wrapper
-            self.model = model = FSDP(
-                model,
-                auto_wrap_policy=auto_wrap_policy,
-                auto_wrapper_callable=auto_wrapper_callable,
-                **fsdp_kwargs,
-            )
+            if self.is_fsdp_xla_v2_enabled:
+                # Should we have this logic into the FSDPv2 wrapper?
+                model = model.to(xm.xla_device())
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply src/transformers/trainer.py:1435]: Fixed.

File: src/transformers/trainer.py (hunk 1 above)
Comment (alanwaketan) [reply src/transformers/trainer.py:1396]: May I ask what's the benefits of doing so?

File: src/transformers/trainer.py (hunk 2 above)
Comment (alanwaketan) [reply src/transformers/trainer.py:1452]: shard_output is not used by FSDPv1. Shouldn't we guard that with the flag too?

**Comment by alanwaketan** [comment]
> Tests should be added in the `tests/trainer/test_trainer.py` file. You should find similar tests!

Speaking of adding tests, what should I test? I mean do you have TPU CI?

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -1961,6 +2003,11 @@ def _inner_training_loop(
                     self.control = self.callback_handler.on_substep_end(args, self.state, self.control)
 
                 if self.control.should_epoch_stop or self.control.should_training_stop:
+                    # PyTorch/XLA relies on the data loader to insert the mark_step for
+                    # each step. Since we are breaking the loop early, we need to manually
+                    # insert the mark_step here.
+                    if is_torch_tpu_available():
```
Comment (alanwaketan) [inline src/transformers/trainer.py:2009]: I fixed a bug here. cc @ArthurZucker @jonb377

**Comment by alanwaketan** [comment]
The test failures don't seem to be related. I tried rebasing as well.

**Review by muellerzr — APPROVED** [review]

**Comment by alanwaketan** [comment]
Thanks @ArthurZucker and @muellerzr for approving the change.

**Comment by alanwaketan** [comment]
It's all green. Can HF folks help with landing the PR? Appreciate it.

**Comment by amyeroberts** [comment]
I can merge :) Thanks for adding this support @alanwaketan!

**Pull request merged**